File: sdk/dart/dovahlink_client/lib/src/protocol/state_snapshot_payload.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:dovahlink_client_sdk/src/protocol/json_map.dart';
import 'package:dovahlink_client_sdk/src/protocol/protocol_format_exception.dart';
import 'package:dovahlink_client_sdk/src/protocol/protocol_revision_validator.dart';
import 'package:dovahlink_client_sdk/src/protocol/protocol_timestamp_validator.dart';

part 'state_snapshot_payload.g.dart';

/// Incoming `state_snapshot` payload (`protocol/schema/README.md`'s `state_snapshot`). Decode-only:
/// the client never sends `state_snapshot`. `data` stays an untyped [JsonMap]: no state area is
/// currently registered (protocol/schema/README.md's "Registered state areas"), so there is no
/// domain shape to decode it against yet.
@JsonSerializable(checked: true, createToJson: false)
class StateSnapshotPayload {
  /// The state area this snapshot represents.
  @JsonKey(required: true)
  final String stateArea;

  /// The revision established by this snapshot.
  @JsonKey(required: true)
  final int revision;

  /// UTC RFC 3339 wall-clock time for display and diagnostics; not an ordering source.
  @JsonKey(required: true)
  final String occurredAt;

  /// The state-area-specific snapshot data.
  @JsonKey(required: true)
  final JsonMap data;

  /// Creates a state-snapshot payload.
  const StateSnapshotPayload({
    required this.stateArea,
    required this.revision,
    required this.occurredAt,
    required this.data,
  });

  /// Decodes and validates one `state_snapshot` payload.
  factory StateSnapshotPayload.fromJson(JsonMap json) {
    try {
      ProtocolRevisionValidator.validateNonNegativeInt(
        json['revision'],
        'revision',
      );
      final StateSnapshotPayload payload = _$StateSnapshotPayloadFromJson(json);
      ProtocolTimestampValidator.validateUtcRfc3339(payload.occurredAt);
      return payload;
    } on Object catch (error) {
      throw ProtocolFormatException('Invalid state_snapshot payload: $error');
    }
  }
}

File: integration/DovahLinkValidationClient/StateSnapshotPayload.cs
using System.Text.Json.Nodes;

namespace DovahLinkValidationClient;

/// <summary>
/// Incoming <c>state_snapshot</c> payload (<c>protocol/schema/README.md</c>'s <c>state_snapshot</c>).
/// Decode-only: this client never sends <c>state_snapshot</c>. <c>Data</c> stays an untyped
/// <see cref="JsonObject"/>: no state area is currently registered (protocol/schema/README.md's
/// "Registered state areas"), so there is no domain shape to decode it against yet.
/// </summary>
/// <param name="StateArea">The state area this snapshot represents.</param>
/// <param name="Revision">The revision established by this snapshot.</param>
/// <param name="OccurredAt">UTC RFC 3339 wall-clock time for display and diagnostics; not an
/// ordering source.</param>
/// <param name="Data">The state-area-specific snapshot data.</param>
public sealed record StateSnapshotPayload(string StateArea, int Revision, string OccurredAt, JsonObject Data)
{
    /// <summary>
    /// Decodes and validates one <c>state_snapshot</c> payload.
    /// </summary>
    /// <param name="payload">The envelope's decoded payload object.</param>
    /// <returns>The decoded state-snapshot payload.</returns>
    /// <exception cref="FormatException">Thrown when a required field is missing, the wrong JSON
    /// type, or fails semantic validation (an empty state area, negative revision, or malformed
    /// timestamp).</exception>
    public static StateSnapshotPayload Decode(JsonObject payload)
    {
        try
        {
            string stateArea = payload["stateArea"]?.GetValue<string>()
                ?? throw new FormatException("Missing stateArea.");
            if (stateArea.Length == 0)
            {
                throw new FormatException("stateArea must be a non-empty string.");
            }
            int revision = payload["revision"]?.GetValue<int>()
                ?? throw new FormatException("Missing revision.");
            ProtocolRevisionValidator.ValidateNonNegative(revision, "revision");
            string occurredAt = payload["occurredAt"]?.GetValue<string>()
                ?? throw new FormatException("Missing occurredAt.");
            ProtocolTimestampValidator.ValidateUtcRfc3339(occurredAt);
            JsonObject data = payload["data"] as JsonObject ?? throw new FormatException("Missing or malformed data.");
            return new StateSnapshotPayload(stateArea, revision, occurredAt, data);
        }
        catch (InvalidOperationException ex)
        {
            throw new FormatException($"Malformed state_snapshot payload: {ex.Message}", ex);
        }
    }
}
